ldms/src/sampler/ibmad_sampler/ibmad_sampler.c

```diff
@@ -523,10 +602,108 @@ static void metrics_tree_sample()
 	memcpy(&tv_prev, &tv_now, sizeof(tv_prev));
 }
 
+static void reinit_ports()
+{
+	int i;
+	conf.port_filter = PORT_FILTER_NONE;
+	for (i = 0; i < MAX_CA_NAMES; i++) {
+		conf.ports[i].ca_name[0] = '\0';
+		conf.ports[i].port_bits = 0;
+	}
+}
+
+static void dump_port_filters()
+{
+	int i;
+	log_fn(LDMSD_LDEBUG, SAMP ": dump_port_filters: filt=%s\n",
+		(conf.port_filter == PORT_FILTER_NONE ? "NONE" : (
+			conf.port_filter == PORT_FILTER_INCLUDE ?
+				"INCLUDE" : "EXCLUDE")));
+	for (i = 0 ; i < MAX_CA_NAMES; i++) {
+		if (conf.ports[i].ca_name[0] == '\0')
+			break;
+		log_fn(LDMSD_LDEBUG, SAMP ": dpf: %s : 0x%x\n",
+			conf.ports[i].ca_name, conf.ports[i].port_bits);
+	}
+}
+
+static int parse_port_filters(const char *include, const char *exclude)
+{
+	if (include && exclude) {
+                log_fn(LDMSD_LERROR, SAMP": config: specify either include or exclude option but not both.\n");
+		return 1;
+	}
+	const char *val = NULL;
+	if (include) {
+		val = include;
+		conf.port_filter = PORT_FILTER_INCLUDE;
+	}
+	if (exclude) {
+		val = exclude;
+		conf.port_filter = PORT_FILTER_EXCLUDE;
+	}
+	int k = 0;
+	int num_ca = 0;
+	if (val) {
+		uint64_t num;
```

Comment:
strtoull may be bigger than unsigned long. but since the documentation and code call out 63 as the upper limit, i will switch types here .  I reflexively use u64 for almost everything in ldms, since it least often bites us later.